### PR TITLE
config refactoring

### LINKED ISSUE #52: ULC config refactoring
# Problem
geth allows using toml config for all params in toml file(geth --config config.toml).
But we store ulc config only in custom json config. 

# Implementation
Change configuration of ulc which should use config by the generic way.
 - allow toml config for ulc
 - command line params for trusted nodes and minTrustedFraction

# Acceptance Criteria
* We can run geth --config config.toml with ulc params
* We can run geth flags for trusted nodes and minTrustedFraction

### PR DESCRIPTION
**Pull request by JekaMas**
closes #52 

should be merged into "ulc/master"

### CODE REVIEW

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -166,13 +166,23 @@ var (
...
 	}
 	ULCModeConfigFlag = cli.StringFlag{
-		Name:  "ulcconfig",
+		Name:  "les.ulcconfig",
 		Usage: "Config file to use for ULC mode",
 	}
 	OnlyAnnounceModeFlag = cli.BoolFlag{
-		Name:  "onlyannounce",
+		Name:  "les.onlyannounce",
 		Usage: "Les server sends only announce",
```
Comment (themue) [inline cmd/utils/flags.go]: Does not belong to this PR but always get troubles with "Les" instead of "LES". It is an acronym like "ULC. See https://github.com/golang/go/wiki/CodeReviewComments#initialisms.

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -737,24 +747,37 @@ func setIPC(ctx *cli.Context, cfg *node.Config) {
 	}
 }
 
-// SetULCFromFile setup ULC config from file if given.
-func SetULCFromFile(ctx *cli.Context, cfg *eth.Config, p2pCfg *p2p.Config) {
-	path := ctx.GlobalString(ULCModeConfigFlag.Name)
-	if path == "" {
+// SetULC setup ULC config from file if given.
+func SetULC(ctx *cli.Context, cfg *eth.Config) {
+	//ulc config haven't loaded from global config, ULC Config and ULC Trusted Nodes not defined
```
Comment (themue) [inline cmd/utils/flags.go]: Comments are code too, for later maintainers. Always worth proper accuracy.

```
// ULC config isn't loaded from global config and ULC config and ULC trusted nodes are not defined.
```

File: cmd/utils/flags.go [hunk 3]
```diff
@@ -737,24 +747,37 @@ func setIPC(ctx *cli.Context, cfg *node.Config) {
...
+	if cfg.ULC==nil && (ctx.GlobalIsSet(ULCModeConfigFlag.Name) || ctx.GlobalIsSet(ULCTrustedNodesFlag.Name)) == false  {
 		return
 	}
+	cfg.ULC=&eth.ULCConfig{}
```
Comment (themue) [inline cmd/utils/flags.go]: No spaces around assignment sign? Looks like missing `gofmt`.

File: cmd/utils/flags.go [hunk 4]
```diff
@@ -737,24 +747,37 @@ func setIPC(ctx *cli.Context, cfg *node.Config) {
...
+	path := ctx.GlobalString(ULCModeConfigFlag.Name)
+	if path != "" {
+		cfgData, err := ioutil.ReadFile(path)
+		if err != nil {
+			Fatalf("Failed to read ULC config file: %v", err)
+		}
+
+		err = json.Unmarshal(cfgData, &cfg.ULC)
+		if err != nil {
+			Fatalf(err.Error())
```
Comment (themue) [inline cmd/utils/flags.go]: Worth a helping string like above. _"Failed to unmarshal ULC configuration: %v"_

File: cmd/utils/flags.go [hunk 5]
```diff
@@ -737,24 +747,37 @@ func setIPC(ctx *cli.Context, cfg *node.Config) {
...
+		}
 	}
 
-	err = json.Unmarshal(cfgData, &cfg.ULC)
-	if err != nil {
-		Fatalf(err.Error())
+	if trustedNodes:=ctx.GlobalString(ULCTrustedNodesFlag.Name); trustedNodes!="" {
```
Comment (themue) [inline cmd/utils/flags.go]: Once again missing spaces. So `gofmt`.

File: cmd/utils/flags.go [hunk 6]
```diff
@@ -737,24 +747,37 @@ func setIPC(ctx *cli.Context, cfg *node.Config) {
...
+		cfg.ULC.ULCTrustedNodes=strings.Split(trustedNodes,";")
```
Comment (themue) [inline cmd/utils/flags.go]: Could it be that nodes passed by flag could be added to those passed by configuration file before instead replacing them? Also most flags split lists with a comma. Any reason to use a semicolon here?

File: cmd/utils/flags.go [hunk 7]
```diff
@@ -737,24 +747,37 @@ func setIPC(ctx *cli.Context, cfg *node.Config) {
...
 	}
 
-	eth.SetULC(cfg.ULC, p2pCfg)
+	if trustedFraction:=ctx.GlobalInt(ULCMinTrustedFractionFlag.Name); trustedFraction>0 {
+		cfg.ULC.MinTrustedFraction=trustedFraction
+	}
+	if cfg.ULC.MinTrustedFraction<=0 && cfg.ULC.MinTrustedFraction> 100 {
```
Comment (themue) [inline cmd/utils/flags.go]: Usage of illegal fraction should be logged, too.

File: eth/ulc_config.go [hunk 8]
```diff
@@ -1,36 +1,9 @@
...
-	"github.com/ethereum/go-ethereum/p2p/discover"
-)
-
-const defaultUTCMinTrustedFraction = 75
+const DefaultUTCMinTrustedFraction = 75
 
 // ULCConfig is a Ultra Light client options.
 type ULCConfig struct {
-	TrustedNodes       []string `toml:",omitempty"` // A list of trusted servers
+	ULCTrustedNodes    []string `toml:",omitempty"` // A list of trusted servers
```
Comment (themue) [inline eth/ulc_config.go]: Why prefix `ULC` here? We are already inside the `ULCConfig`.

File: les/backend.go [hunk 9]
```diff
@@ -111,8 +111,12 @@ func New(ctx *node.ServiceContext, config *eth.Config) (*LightEthereum, error) {
 		bloomTrieIndexer: light.NewBloomTrieIndexer(chainDb, true),
 	}
 
+	var trustedNodes []string
+	if leth.config.ULC != nil {
+		trustedNodes = leth.config.ULC.ULCTrustedNodes
```
Comment (themue) [inline les/backend.go]: See `ULC` prefix here, reads a bit strange.

File: les/serverpool.go [hunk 10]
```diff
@@ -104,6 +104,7 @@ type serverPool struct {
 	discNodes     chan *discv5.Node
 	discLookups   chan bool
 
+	trustedNodes []string
```
Comment (themue) [inline les/serverpool.go]: Missing indentation, so no `gofmt`. New editor or troubles with configuration?

File: les/serverpool.go [hunk 11]
```diff
@@ -411,6 +413,20 @@ func (pool *serverPool) loadNodes() {
 
 }
 
+func (pool *serverPool) parseULCTrustedNodes() []*discover.Node {
```
Comment (themue) [inline les/serverpool.go]: Missing a helpful comment.

```
// parseULCTrustedNodes ...
```

File: eth/ulc_config.go (hunk 8 above)
Comment (b00ris) [reply eth/ulc_config.go]: I think it's because there is trusted nodes field in p2p.Server and p2p.Config. There is a lot of embedding of these structures.

File: cmd/utils/flags.go (hunk 4 above)
Comment (JekaMas) [reply cmd/utils/flags.go]: Ok, will copy-paste :)

File: eth/ulc_config.go (hunk 8 above)
Comment (JekaMas) [reply eth/ulc_config.go]: Maybe just `TrustedServers`?

**Comment by JekaMas** [comment]
@themue I've renamed config field and methods with ULC prefix.
